bundles/org.eclipse.osgi/felix/src/org/apache/felix/resolver/Logger.java

```diff
@@ -131,4 +132,29 @@ public void logUsesConstraintViolation(Resource resource, ResolutionError error)
     {
         // do nothing by default
     }
+
+	/**
+	 * Called whenever a new permutation is added by the resolver.
+	 * 
+	 * @param type      the type of the permutation
+	 * @param remaining a function that can be used to query the now current number
+	 *                  of permutation types
+	 */
+	public void logPermutationAdded(PermutationType type, ToIntFunction<PermutationType> remaining) 
```

Comment:
Using a function here seems like overkill vs simply passing in the calculated integer.

Reply:
My idea here was that for the noop implementation this means the numbers are not calculated, its a simple size() call so probably not that of an impact but something I considered here.
The second thing would be that the method needs to be enhanced once there is a different permutation type maybe in the future, while with this approach the method can stay the same.